### PR TITLE
Post Authentication

### PR DESCRIPTION
**Pull request by emileswarts**
Implement functionality of current backend when dealing with PostAuth.

It creates session records, updates userdetails, and emits different http status codes based on certain conditions.  

There is still a bit of work to do around edge cases like when some values are not present in the URL (which happens frequently).  This will be addressed in the next PR, along with refactoring of the production code.

### CODE REVIEW

File: lib/user.rb [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class User < Sequel::Model(:userdetails)
+  User.unrestrict_primary_key
```
Comment (naliwajek) [inline lib/user.rb]: Why, btw?

Comment (emileswarts) [reply lib/user.rb]: The primary key for userdetails is username.
Without this setting, we won't be able to set it explicitly.

https://sequel.jeremyevans.net/rdoc/classes/Sequel/Model/ClassMethods.html#method-i-unrestrict_primary_key

File: spec/features/post_auth_spec.rb [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            post_auth_request
+            expect(user.last_login).to be_nil
+          end
+
+          it 'returns a 204 status code' do
+            expect(last_response.status).to eq(204)
+          end
+
+          it 'does not create a session record' do
+            expect(Session.all.count).to eq(0)
```
Comment (adrianclay) [inline spec/features/post_auth_spec.rb]: You aren't using the ` .all ` in the it block above.

File: spec/features/post_auth_spec.rb [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          it 'does updates the last login' do
+            post_auth_request
+            expect(user.last_login).to_not be_nil
+          end
+        end
+      end
+
+      context 'MAC Formatter' do
+        let(:mac) { '50a67f849cd1' }
+        it 'calls the MAC Formatter' do
```
Comment (adrianclay) [inline spec/features/post_auth_spec.rb]: ` it 'saves the mac formatted' `?

File: spec/features/post_auth_spec.rb (hunk 2 above)
Comment (emileswarts) [reply spec/features/post_auth_spec.rb]: Nice catch, fixed

File: spec/features/post_auth_spec.rb [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          expect(Session.count).to eq(0)
+          expect(last_response.status).to eq(404)
+        end
+      end
+
+      context 'Blank authentication result' do
+        let(:authentication_result) { '' }
+
+        it 'returns a 404 for anything other than Access-Accept or Access-Reject' do
+          expect(last_response.status).to eq(404)
```
Comment (adrianclay) [inline spec/features/post_auth_spec.rb]: Is this behaviour important?  Why do we need this to return a 404 in this case?

File: app.rb [hunk 5]
```diff
@@ -1,9 +1,35 @@
...
+        mac = MacFormatter.new.execute(mac: params.fetch(:mac))
+        Session.create(
+          username: params.fetch(:username),
+          mac: mac,
+          ap: params.fetch(:called_station_id),
+          siteIP: params.fetch(:site_ip_address),
+          building_identifier: params.fetch(:called_station_id),
+        )
+        user = User.find(username: params.fetch(:username))
+        user.last_login = Time.now.strftime('%y-%m-%d %H:%M:%S')
```
Comment (adrianclay) [inline app.rb]: What were your thoughts on removing the coupling between the user table and this service?

File: spec/features/post_auth_spec.rb (hunk 4 above)
Comment (emileswarts) [reply spec/features/post_auth_spec.rb]: I don't see harm in being explicit about this.
Testing that 204 status codes are returned from Access-Accept and Access-Reject implies that anything else would be a 404

File: app.rb (hunk 5 above)
Comment (emileswarts) [reply app.rb]: Only reasonable option that comes to mind is to expose a update-last-login endpoint on the user-signup API.  We call that from here.

Comment (adrianclay) [reply app.rb]: Sounds sensible to me.  Feel as though we can introduce that as a separate slice in this iteration.

Comment (emileswarts) [reply app.rb]: Agreed

File: spec/features/post_auth_spec.rb (hunk 4 above)
Comment (adrianclay) [reply spec/features/post_auth_spec.rb]: Are we expecting to see this 404 behaviour in production?

**Pull request merged**